Add saved questions picker

alxnddr opened the pull request:
### Description
Add saved question picker that shows collections tree.

#### Some important facts
- In custom mode when a table is selected and users join a saved question the picker shows saved questions only from the already selected database.
- When the saved question picker is active the search shows only saved questions.

### Screenshot
<img width="500" alt="Screen Shot 2021-06-30 at 21 14 38" src="https://user-images.githubusercontent.com/14301985/124011044-3f259600-d9e8-11eb-842a-65886b38a226.png">

This is rework of the previous PR #15663 (which itself was the rework of PR #14864)

alxnddr committed:
Add Tree component

alxnddr committed:
Add Saved Question picker

alxnddr committed:
fix tree spec

Commit:
Basic sanity E2E tests

ariya commented on frontend/src/metabase/query_builder/components/DataSelector.jsx:
```diff
@@ -705,12 +725,19 @@ export class UnconnectedDataSelector extends Component {
     const isSearchActive = searchText.trim().length >= MIN_SEARCH_LENGTH;
 
+    const searchPlaceholder = isSavedQuestionPickerShown
+      ? `Search for a question...`
```
Just a note: I'm unsure if this adds a new string for the translator (late in the process), but it turns out that we have already a similar string "Search for a question" (without the triple dots). Either we use that (however, it'll be inconsistent with "Searching for a table...") or we use the generic "Search" (this string is already everywhere).
Alternatively, we just keep it as is (with the triple dots) and ask the translators to catch up in 40.1 or something.

mazameli replied:
If this is going in 0.40, we should use the already-translated string, please.

alxnddr replied:
I used the existing "Search for a question"

alxnddr committed:
Merge remote-tracking branch 'origin/master' into feat/saved-question-picker

alxnddr committed:
fix bugs, minor refactoring

alxnddr committed:
keep item selected in saved question picker

ariya commented on frontend/src/metabase/query_builder/components/data-search/utils.js:
```diff
@@ -1,12 +1,22 @@
   return searchResultItem;
 }
+
+export function isSavedQuestion(table) {
```
Where is this function being used?

alxnddr replied:
Nice catch, removed it

Commit:
E2E tests for scoped search

ariya commented:
Overall, it looks super good :+1: .

There's a minor issue (likely because the state `isSavedQuestionPickerShown` isn't initialized properly) if an existing question is edited. Try New Custom Question, Saved Questions, choose any question and then Visualize. Then click the Show Editor to enter the notebook again and click on the question name. This would start over at Sample Dataset (should be showing saved question picker instead).

However, looks like it's only a minor annoyance and I'm fine if we address it as a follow-up (not necessarily lumped in this PR).

alxnddr committed:
adjust tree item icon color, prevent saved question list flickering

alxnddr committed:
remove unused function

alxnddr commented:
@ariya thanks for catching this, I'll create a follow-up PR

kulyk commented on frontend/src/metabase/query_builder/components/data-search/SearchResults.jsx:
```diff
@@ -9,10 +9,22 @@ import Search from "metabase/entities/search";
+  databaseId: PropTypes.string,
+  searchQuery: PropTypes.string.required,
+  onSelect: PropTypes.func.required,
+  searchModels: PropTypes.arrayOf(PropTypes.string.isRequired),
```
nip: should we maybe list the `searchModels` options right away if we know what's in use now?
Like `PropTypes.arrayOf(PropTypes.oneOf(["card", "table"]))`

kulyk commented on frontend/src/metabase/query_builder/components/saved-question-picker/SavedQuestionPicker.jsx:
```diff
@@ -0,0 +1,104 @@
+  tableId: PropTypes.string,
+};
+
+const OUR_ANALYTICS_COLLECTION = {
+  id: "root",
+  schemaName: "Everything else",
+  name: t`Our analytics`,
+  icon: "folder",
+};
```
Should we maybe use the `ROOT_COLLECTION` object from [here](https://github.com/metabase/metabase/blob/ab50eaaa2fdf0583092f99ee11e8fc9a298582bf/frontend/src/metabase/entities/collections.js#L18), so we have a single source of truth on root ID and name?

ariya reviewed (approved):
Excellent work. Let's ship it :1st_place_medal:

kulyk reviewed (commented):
Looks great, both the component and the code 👏

Noticed two a bit confusing behaviors. Should we address them or is it by design / out of scope of this PR?

### Collections tree sync

1. Let's say all my questions are in "Our analytics"
2. Move a question in any other collection, let's say to a personal one
3. Click "Ask a question" in the navigation bar > Simple Question > Saved Questions

**Expected**: the collection tree displays both "Our analytics" and my personal collection

**Actual**: the collection tree displays "Our analytics" only. However, the moved question is still there

**Note**: if you refresh the page, the tree is shown as expected

### Questions in nested collections

1. Let's say all my questions are in "Our analytics"
2. Create collections "C1" and "C2", "C2" has to be a child of "C1"
3. Move a question to "C2", "C1" has to be empty
4. Click "Ask a question" in the navigation bar > Simple Question
5. Refresh the page to avoid the previous issue
6. Select "Saved Questions"

**Expected**: there is either a "C2" in the collection tree or "C1" > "C2" tree

**Actual**: both "C1" and "C2" are not displayed, so the only way to access a question from "C2" is to use search

### Demo

I recorded a 2min video (with voiceover) where I reproduced both issues if my descriptions are unclear


https://user-images.githubusercontent.com/17258145/124159579-ff23e900-daa3-11eb-91ad-f789abd22038.mov

ariya commented:
> **Note**: if you refresh the page, the tree is shown as expected

I believe this is a generic/known state sync issue (way before this PR). Try this: create a simple question and save it. Then start another question and try to find that new saved question. You won't immediately find it, until you refresh the page.

alxnddr committed:
fix incorrectly hidden collections in the saved question picker

alxnddr committed:
change icon for personal collections

alxnddr committed:
make collection tree and question list scroll bars independent

alxnddr committed:
review fixes

alxnddr commented:
@kulyk thanks for your review! You caught the bug with hidden collections and I've fixed it now and addressed other suggestions. Regarding the tree sync without page reload I think we should address this problem separately because it works this way in the entire app
